### ISSUE TITLE
New rule: Flag functions that could be changed reified function

### ISSUE THREAD

**Issue by 3flex** [arturbosch/detekt#1881]
## Expected Behavior of the rule
Flag functions that take a class type as a parameter and suggest converting to a reified function.

Example from detekt's code base:
```kotlin
fun PsiElement.isPartOf(clazz: KClass<out PsiElement>): Boolean = 
        getNonStrictParentOfType(clazz.java) != null
```

This is better as:

```kotlin
inline fun <reified T : PsiElement> PsiElement.isPartOf(): Boolean =
         getNonStrictParentOfType(T::class.java) != null
```

## Context
Usage at call sites is nicer - instead of `isPartOf(KtStringTemplateEntry::class)` they can use `isPartOf<KtStringTemplateEntry>()`

**Comment by schalkms** [comment]
I like this idea. Developers switching from Java might not be aware of this language feature in Kotlin.
We should note that an inlined function with a reified type is not callable from Java code, though.

**Comment by TWiStErRob** [comment]
Note: it's not a silver bullet; it's also not possible to calculate a variable and use it when the function is reified:
```kotlin
val cls = if (foo) bar else baz
if (element.isPartOf<cls>()) ...
```

**Comment by schalkms** [comment]
Thanks for mentioning that! @TWiStErRob 
Yes, there are some limitations. This rule would require intensive testing.
Anyhow, no static analysis rule is immune against false positives.
